client-cli/src/main/java/au/org/consumerdatastandards/client/cli/support/ApiUtil.java

```diff
@@ -111,21 +108,37 @@ private static X509Certificate loadCertificate(String certFilePath) throws Certi
     private static PrivateKey loadPrivateKey(String keyFilePath)
         throws IOException, ApiException, NoSuchAlgorithmException, InvalidKeySpecException {
         Security.addProvider(new BouncyCastleProvider());
-        final String START_OF_BEGIN = "-----BEGIN ";
-        final String END_OF_BEGIN = " PRIVATE KEY-----";
-        String content = new String(Files.readAllBytes(Paths.get(keyFilePath)));
-        String[] lines = content.split("\\R");
-        if(!lines[0].startsWith(START_OF_BEGIN) || !lines[0].endsWith(END_OF_BEGIN)) {
+        FileReader reader = new FileReader(keyFilePath);
+        PemReader pemReader = new PemReader(reader);
+        PemObject pemObject = pemReader.readPemObject();
+        String type = pemObject.getType();
+        final String PRIVATE_KEY_TYPE_SUFFIX = "PRIVATE KEY";
+        if (!type.endsWith(PRIVATE_KEY_TYPE_SUFFIX)) {
             throw new ApiException("Invalid key file content - expecting first line similar to\n" +
                 "-----BEGIN RSA PRIVATE KEY-----");
         }
-        String algorithmName = lines[0].replace(START_OF_BEGIN, "").replace(END_OF_BEGIN, "");
-        String lastLine = lines[lines.length - 1];
-        String encoded = content.replace(lines[0] + "\n", "").replace(lastLine, "");
-        byte[] pkcs8Encoded = Base64.decodeBase64(encoded);
-        PKCS8EncodedKeySpec keySpec = new PKCS8EncodedKeySpec(pkcs8Encoded);
+        PKCS8EncodedKeySpec keySpec = new PKCS8EncodedKeySpec(pemObject.getContent());
+        String algorithmName = type.replace(PRIVATE_KEY_TYPE_SUFFIX, "").trim();
+        if (StringUtils.isNotBlank(algorithmName)){
+            KeyFactory kf = KeyFactory.getInstance(algorithmName);
+            return kf.generatePrivate(keySpec);
+        } else {
+             String[] algos = {"RSA", "DSA", "EC"};
+             for (String algo: algos) {
+                 PrivateKey key = tryGeneratingPrivateKey(keySpec, algo);
+                 if (key != null) return key;
+             }
+             throw new ApiException("Unsupported public key algorithm");
```

Comment:
Let tryGeneratingPrivateKey() throw the exception instead of null, handle it in the loop, but attach it as a cause here, just in case it was thrown because of the completely messed-up key.

Reply:
updated